packages/server/src/plugin/subscriptionCallback/index.ts

```diff
@@ -0,0 +1,655 @@
+      // If it's not a callback subscription, ignore the request.
+      if (!request?.extensions?.subscription) return;
+
+      return {
+        // Implementing `responseForOperation` is the only hook that allows us
+        // to bypass normal execution by returning our own response. We don't
+        // want Apollo Server to actually handle this subscription request, we
+        // want to handle everything ourselves. The actual subscription handling
+        // will be done in `willSendResponse`. The router expects the initial
+        // response to be a 200, with the `subscription-protocol` header set to
+        // `callback`.
+        async responseForOperation({ request }) {
+          logger?.debug(
+            'Received new subscription request',
+            request.extensions!.subscription.subscription_id,
+          );
+
+          return {
+            http: {
+              status: 200,
+              headers: new HeaderMap([['subscription-protocol', 'callback']]),
+            },
+            body: {
+              kind: 'single',
+              singleResult: {},
+            },
+          };
+        },
+        // The majority of the subscription work is implemented in this hook.
+        // This can _almost_ all be implemented in `responseForOperation`,
+        // however in the case of GraphQL validation errors,
+        // `responseForOperation` won't actually be called, but
+        // `willSendResponse` will.
+        async willSendResponse({
+          request,
+          schema,
+          document,
+          contextValue,
+          operationName,
+          response,
+        }) {
+          const {
+            callback_url: callbackUrl,
+            subscription_id: id,
+            verifier,
+          } = request.extensions!.subscription;
+          try {
+            logger?.debug(
+              'Sending `check` request to router',
+              request.extensions!.subscription.subscription_id,
+            );
+            // Before responding to the original request, we need to complete a
+            // roundtrip `check` request to the router, so we `await` this
+            // request.
+            await subscriptionManager.checkRequest({
+              callbackUrl,
+              id,
+              verifier,
+            });
+            logger?.debug(
+              '`check` request successful',
+              request.extensions!.subscription.subscription_id,
+            );
+          } catch (e) {
+            const graphqlError = ensureGraphQLError(e);
+            logger?.error(
+              `\`check\` request failed: ${graphqlError.message}`,
+              id,
+            );
+            // In the event of a check failure, we respond to the original
+            // request with a >=400 status code.
+            if (response.body.kind === 'single') {
+              response.body.singleResult.errors = [graphqlError];
+              response.http.status = 400;
```

Comment:
Why not a 5xx? Is this something we expect to happen in a non-broken case? (Also maybe we should pay attention to the http extension on graphqlError?)

Reply:
Good catch, agree this should be 500. I'm not sure the `graphqlError.extensions.http.status` is relevant here, but maybe you have something in mind? The error being caught here is coming directly from `fetch` or being thrown by us, and per the protocol it needs to be >=400. I could do a `Math.max()` to make sure the status is at least 400 if a status is given but that seems a bit weird.

Reply:
[29ab4f1](https://github.com/apollographql/apollo-server/pull/7617/commits/29ab4f1032b0c203646b5f747a8c2c0c8c6bf4e1)